include/rmm/mr/device/statistics_resource_adaptor.hpp

```diff
@@ -145,7 +173,40 @@ class statistics_resource_adaptor final : public device_memory_resource {
   {
     read_lock_t lock(mtx_);
 
-    return allocations_;
+    return counter_stack_.top().second;
+  }
+
+  /**
+   * @brief Push a pair of zero counters on the stack, which becomes the new
+   * counters returned by `get_bytes_counter()` and `get_allocations_counter()`
+   *
+   * @return top pair of counters <bytes, allocations> from the stack _before_
+   * the push
+   */
+  std::pair<counter, counter> push_counters()
+  {
+    write_lock_t lock(mtx_);
```

Comment:
It's weird to think of this stack being used in a multithreaded situation. We have the mutexes, but two threads could possibly push the stack from two completely different functions. I guess that is just a "shoot yourself in the foot" situation that users would hopefully avoid?

Reply:
Agree, and even if two threads are pushing to the stack, the work done within the locked region is very fast in most cases. It is only when `counter_stack_` needs to grow that it will take a bit longer.

the user

Reply:
Perhaps the stack should be thread local. Otherwise the pushed/popped statistics won't make sense.